clearwater-queue-manager.root/usr/share/clearwater/clearwater-queue-manager/scripts/upgrade_cassandra.py

```diff
@@ -0,0 +1,204 @@
+# Not sure if we need the non-CC packages here
+PACKAGES = "cassandra clearwater-cassandra homestead-cassandra homestead-prov-cassandra homer-cassandra memento-cassandra"
+SOURCES_FILE = "/etc/apt/sources.list.d/clearwater.list"
+TEMP_SOURCES_FILE = "/tmp/clearwater.list"
+VERSION_FILE = "/etc/clearwater/cassandra_version"
+
+_log = logging.getLogger("cassandra_upgrade_plugin")
+
+def get_cassandra_version():
+    version = ""
+    nodetool_process = subprocess.Popen("/usr/share/clearwater/bin/run-in-signaling-namespace nodetool version",
+                                        stdout=subprocess.PIPE,
+                                        shell=True)
+
+    # The output of nodetool version is like "ReleaseVersion: 2.0.14"
+    version_regex = "ReleaseVersion: (.*)$"
+    for line in nodetool_process.stdout:
+        line = line.strip()
+        m = re.match(version_regex, line)
+        if m:
+            version = m.group(1)
+            break
+
+    return version
+
+# Attempts the Cassandra upgrade.
+# Returns True if the upgrade succeeded or no upgrade was necessary.
+# Returns False if not.
+def do_upgrade():
+    # First, read the desired cassandra version and check we need to upgrade
+    new_version = ""
+    repo_dir = ""
+    with open(VERSION_FILE, "r") as version_file:
+        for line in version_file:
+            line = line.strip()
+            if line.startswith("desired_version"):
+                # The version is specified as desired_version=CCxxxxx_cass<cass_version>
+                split_line = line.split("=")
+                if len(split_line) == 2:
+                    #  The repo_dir is everything after the equals
+                    repo_dir = split_line[1].strip()
+
+                    # The cassandra version is everything after the "cass"
+                    split_repo_dir = repo_dir.split("cass")
+                    if len(split_repo_dir) == 2:
+                        new_version = split_repo_dir[1].strip()
+            break
+
+    if new_version == "":
+        _log.error("Unable to read desired Cassandra version from {0}".format(VERSION_FILE))
+        return False
+
+    old_version = get_cassandra_version()
+    if old_version == "":
+        _log.error("Unable to get current Cassandra version")
+        return False
+
+    if new_version == old_version:
+        _log.info("Current version equal to desired version: {0}; nothing to do".format(new_version))
+        return True
+
+    _log.info("Upgrading Cassandra from version {0} to version {1}".format(old_version, new_version))
+
+    if run_command("/usr/share/clearwater/bin/run-in-signaling-namespace nodetool upgradesstables"):
+        _log.error("Unable to upgrade sstables")
+        return False
+
+    _log.info("Updating cassandra repo")
+
+    # Create a backup of the sources list file to roll back to if necessary
+    shutil.copyfile(SOURCES_FILE, TEMP_SOURCES_FILE)
+
+    # The clearwater sources list should include a single source. Update it
```

Comment:
We spoke about sticking a comment in the sources file to make it easier to change this in future - why isn't this here?